src/main/kotlin/org/jitsi/nlj/dtls/DtlsStack.kt

```diff
@@ -13,55 +13,46 @@
-import org.jitsi.rtp.UnparsedPacket
-import org.jitsi.utils.logging2.Logger
+import java.nio.ByteBuffer
+import java.time.Duration
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.TimeUnit
+import kotlin.concurrent.thread
+import kotlin.math.min
 
 /**
- * Represents a single instance of a DTLS stack for a given connection.  This class also acts as the [DatagramTransport]
- * used by the underlying DTLS library in order to send and receive DTLS packets.  Users of this class need to handle
- * passing incoming DTLS packets into the stack, as well as handling packets the stack wants to send out.  The passing
- * of incoming packets is done via calling [DtlsStack.processIncomingProtocolData].  The handling of outgoing packets is done by
- * assigning a handler to the [DtlsStack.onOutgoingProtocolData] member.  Incoming packets may be either control packets
- * (terminated by the stack itself) or app packets which have been sent over DTLS (SCTP packets, for example).  After
- * passing incoming packets to the stack via [DtlsStack.processIncomingProtocolData], any app packets ready for further processing
- * will be returned.  Outgoing packets can be sent via [DtlsStack.sendApplicationData].
+ * A DTLS stack implementation, which can act as either the DTLS server or DTLS client, and can be used
+ * for negotiating the DTLS connection and sending and receiving data over that connection.
  *
- * An example of passing incoming DTLS packets through the stack:
+ * This class does not directly communicate with the network in any way.  Raw DTLS packets must be fed into
+ * this stack via the [processIncomingProtocolData] method.  [incomingDataHandler] will be invoked with the
+ * decrypted, application data sent over the connection.
  *
- *  --> Recv 'dtlsPacket' from the network and pass it into the stack:
- *  val appPackets = dtlsStack.processIncomingProtocolData(listOf(dtlsPacket))
- *  if (appPackets.isNotEmpty()) {
- *    // Process the app packets
- *  }
+ * Data can be sent through the DTLS connection via [sendApplicationData]; it will be encrypted by the stack and
+ * then sent out via the [outgoingDataHandler], which must be set to have the data go anywhere interesting :)
  *
- *  An example of sending app packets out via the DTLS stack:
- *  dtlsStack.onOutgoingData = { outgoingDtlsPacket ->
- *    // Work to send the packets out
- *  }
- *  val dtlsAppPacket = ...
- *  dtlsStack.sendApplicationData(dtlsAppPacket)
+ * [eventHandler] will be invoked when any events occur.
  *
+ * After wiring up all the handlers, to start a connection the stack must be told which role to fullfil: client
+ * or server.  This can be done by calling either the [actAsClient] or [actAsServer] methods.  Once the role has
+ * been set, [start] can be called to start the negotiation.
  */
 class DtlsStack(
     parentLogger: Logger
-) : ProtocolStack, DatagramTransport, NodeStatsProducer {
+) {
```

Comment:
Are `ProtocolStack`, `ProtocolReceiver`, `ProtocolSender` deprecated now?

Reply:
Yes! Forgot to delete them. That will feel good :)